src/representations/feature_manager_block_sparse.hh

```diff
@@ -264,6 +291,204 @@ namespace rascal {
+      auto iter = st.find(k);
+      if (iter != st.end()) {
+        rtn.push_back(k);
+        st.erase(iter);
+      }
+    });
+    return rtn;
+  }
+
+  /**
+   * Matrix multiplication between two block sparse feature matrices
+   */
+  template <typename Precision_t>
+  decltype(auto) dot(FeatureManagerBlockSparse<Precision_t> & X1,
+                     FeatureManagerBlockSparse<Precision_t> & X2) {
+    auto n1_centers{X1.sample_size()};
+    auto n2_centers{X2.sample_size()};
+    Dense_t<Precision_t> mat =
+        Dense_t<Precision_t>::Zero(n1_centers, n2_centers);
+    for (int icenter1{0}; icenter1 < n1_centers; icenter1++) {
+      auto && keys1{X1.get_keys(icenter1)};
+      for (int icenter2{0}; icenter2 < n2_centers; icenter2++) {
+        auto && keys2{X2.get_keys(icenter2)};
+        auto keys_intersection{intersection_of(keys1, keys2)};
+        for (const auto & key : keys_intersection) {
+          auto vec1{X1.get_features(icenter1, key)};
+          auto vec2{X2.get_features(icenter2, key)};
+          // double fac{1.};
+          // if (key[0] != key[1]) {
+          //   fac = 2.;
+          // }
+          mat(icenter1, icenter2) += vec1.dot(vec2);
+        }
+      }
+    }
+    return mat;
+  }
+
+  /**
+   * Matrix multiplication between a block sparse feature matrix with itself
+   */
+  template <typename Precision_t>
+  decltype(auto) dot(FeatureManagerBlockSparse<Precision_t> & X1) {
+    auto n1_centers{X1.sample_size()};
+    Dense_t<Precision_t> mat =
+        Dense_t<Precision_t>::Zero(n1_centers, n1_centers);
+    for (int icenter1{0}; icenter1 < n1_centers; icenter1++) {
+      auto && keys1{X1.get_keys(icenter1)};
+      mat(icenter1, icenter1) = 1.;
+
+      for (int icenter2{icenter1 + 1}; icenter2 < n1_centers; icenter2++) {
+        auto && keys2{X1.get_keys(icenter2)};
+        auto keys_intersection{intersection_of(keys1, keys2)};
+        for (const auto & key : keys_intersection) {
+          auto vec1{X1.get_features(icenter1, key)};
+          auto vec2{X1.get_features(icenter2, key)};
+          mat(icenter1, icenter2) += vec1.dot(vec2);
+        }
+        mat(icenter2, icenter1) = mat(icenter1, icenter2);
+      }
+    }
+    return mat;
+  }
+
+  /**
+   * global cosine kernel with a power of zeta
+   */
+  template <typename Precision_t>
+  decltype(auto)
+  cosine_kernel_global(int zeta, FeatureManagerBlockSparse<Precision_t> & X1,
+                       FeatureManagerBlockSparse<Precision_t> & X2) {
+    auto stride1{X1.get_structure_stride()};
+    auto stride2{X2.get_structure_stride()};
+    auto n1_frames{stride1.size() - 1};
+    auto n2_frames{stride2.size() - 1};
+
+    Dense_t<Precision_t> mat(n1_frames, n2_frames);
+    // loop over the structures of the 1st set
+    for (size_t iframe1{0}; iframe1 < n1_frames; iframe1++) {
+      auto & iframe1_st{stride1[iframe1]};
+      auto & iframe1_nd{stride1[iframe1 + 1]};
```

Comment:
I know that you like to use `nd` for `end`...but it took me again half a minute to figure it out. Can we make `_nd` to `_end`, pleeeeeease :)